lib/ecto/changeset.ex

```diff
@@ -225,33 +225,86 @@ defmodule Ecto.Changeset do
                optional: optional}
   end
 
-  defp process_param(key, kind, params, types, model, {changes, errors}) do
+  defp process_param({key, fun}, kind, params, types, model, acc) do
     {key, param_key} = cast_key(key)
+    type = type!(types, key, fun)
```

Comment:
Actually, if a fun was given, we should certify it is an embed. Otherwise we should raise, so you likely want to have both `type!` and `embed!` functions.